src/ripple/ledger/View.h

```diff
@@ -418,6 +420,647 @@ transferXRP(
+
+    if (finalLockedBalance > balance)
+    {
+        JLOG(j.trace()) << "trustAdjustLockedBalance: "
+                        << "lockedBalance(" << finalLockedBalance
+                        << ") > balance(" << balance << ") = true\n";
+        return tecINSUFFICIENT_FUNDS;
+    }
+
+    if (finalLockedBalance < beast::zero)
+        return tecINTERNAL;
+
+    // check if there is significant precision loss
+    if (!isAddable(balance, deltaAmt) ||
+        !isAddable(priorLockedBalance, deltaAmt) ||
+        !isAddable(finalLockedBalance, balance))
+        return tecPRECISION_LOSS;
+
+    // sanity check possible overflows on the lock counter
+    if ((deltaLockCount > 0 && priorLockCount > finalLockCount) ||
+        (deltaLockCount < 0 && priorLockCount < finalLockCount) ||
+        (deltaLockCount == 0 && priorLockCount != finalLockCount))
+        return tecOVERSIZE;
+
+    // we won't update any SLEs if it is a dry run
+    if (dryRun)
+        return tesSUCCESS;
+
+    if constexpr (
+        std::is_same<V, ApplyView>::value &&
+        std::is_same<S, std::shared_ptr<SLE>>::value)
+    {
+        if (finalLockedBalance == beast::zero || finalLockCount == 0)
+        {
+            sleLine->makeFieldAbsent(sfLockedBalance);
+            sleLine->makeFieldAbsent(sfLockCount);
+        }
+        else
+        {
+            sleLine->setFieldAmount(
+                sfLockedBalance,
+                high ? -finalLockedBalance : finalLockedBalance);
+            sleLine->setFieldU32(sfLockCount, finalLockCount);
+        }
+
+        view.update(sleLine);
+    }
+
+    return tesSUCCESS;
+}
+
+/** Check if a set of accounts can freely exchange the specified token.
+    Read only, does not change any ledger object.
+    May be called with ApplyView or ReadView.
+    (including unlocking) is forbidden by any flag or condition.
+    If parties contains 1 entry then noRipple is not a bar to xfer.
+    If parties contains more than 1 entry then any party with noRipple
+    on issuer side is a bar to xfer.
+*/
+template <class V>
+[[nodiscard]] TER
+trustTransferAllowed(
+    V& view,
+    std::vector<AccountID> const& parties,
+    Issue const& issue,
+    beast::Journal const& j)
+{
+    static_assert(
+        std::is_same<V, ReadView const>::value ||
+        std::is_same<V, ApplyView>::value);
+
+    typedef typename std::conditional<
+        std::is_same<V, ApplyView>::value,
+        std::shared_ptr<SLE>,
+        std::shared_ptr<SLE const>>::type SLEPtr;
+
+    if (isFakeXRP(issue.currency))
+        return tecNO_PERMISSION;
+
+    auto const sleIssuerAcc = view.read(keylet::account(issue.account));
+
+    bool lockedBalanceAllowed =
+        view.rules().enabled(featurePaychanAndEscrowForTokens);
+
+    // missing issuer is always a bar to xfer
+    if (!sleIssuerAcc)
+        return tecNO_ISSUER;
+
+    // issuer global freeze is always a bar to xfer
+    if (isGlobalFrozen(view, issue.account))
```

Comment:
what if the destination account is the issuer account? does it mean the holder is not allowed to transfer?

Reply:
The code you are referencing only requires that the issuer exists. If the issuer is the destination then everything works as normal and on claim the balance is reduced. (Alice has 10k, creates paychan for 1k, gateway claims 500, alice line balance is now 9500)

I had a test suite for this under `testTokenGateway` but it seems the gateway-destination side of the tests have been removed in one of my refactors. I will add those tests back in so that its clear.

Reply:
https://github.com/XRPLF/rippled/pull/4396/commits/44ea1ac953cde3839173d9ffe6490712f040a1fd

Reply:
Sorry if I wasn't clear, I was referring to this code

```
    // issuer global freeze is always a bar to xfer
    if (isGlobalFrozen(view, issue.account))
        return tecFROZEN;
```

Would this mean that even if the destination is the issuer, the holder would still  not be able to transfer back to the issuer if the issuer has set GlobalFreeze?

Reply:
Oh I'm sorry. Yes, the creation of a PayChan/Escrow transactor is rejected if the GlobalFreezeFlag is enabled. 

`NFTokenCreateOffer`, `CreateOffer`, `CreateCheck`, `CashCheck` all look to handle it this way except `Payment`. (There are tests for `accountHolds` that confirm this behavior is possible however I cant seem find where `View_test.cpp` tests this functionality in an e2e test.)

I believe because the other transactors reject the transaction thats why it was written this way. 

Let us know if this is needs to be updated. Maybe we can update all of the transactors (once we double confirm this behavior through tests) in a separate PR.

Reply:
I could only find `isGlobalFrozen` used in `CreateCheck` and `CreateOffer`. But I'm not too sure of the full context of when it should be used. I wonder if @seelabs would know this